src/server/api/contactsApi.js

```diff
@@ -0,0 +1,165 @@
+import { r } from '../models'
+import _ from 'lodash'
+import { getTimezoneByZip, getOptOutSubQuery } from '../../workers/jobs'
+
+import { getValidatedData } from '../../lib'
+
+function campaignaStatusShortCircuit(campaign, res) {
+  let message = ''
+  if (campaign.is_archivedj) {
+    message = 'Campaign is archived'
+  } else if (campaign.is_started) {
+    message = 'Campaign is started'
+  }
+
+  if (message) {
+    res.writeHead(403)
+    res.end(message)
+    return true
+  }
+
+  return false
+}
+
+// TODO(lperson) enforce that this is not reentrant
+// TODO(lperson) enforce auth
+export default async function contactsApi(req, res) {
+  const orgId = req.params.orgId
+  const campaignId = req.params.campaignId
+
+  if (!['GET', 'DELETE', 'POST'].includes(req.method)) {
+    res.writeHead(405, { Allow: 'GET, POST, DELETE' })
+    res.end('Not allowed')
+    return
+  }
+
+  const campaignArray = await r
+    .knex('campaign')
+    .select()
+    .where({ organization_id: orgId, id: campaignId })
+
+  if (campaignArray.length !== 1) {
+    res.writeHead(404)
+    res.end('Not found')
+    return
+  }
+
+  const campaign = campaignArray[0]
+
+  let resp = {}
+
+  switch (req.method) {
+    case 'GET':
+      const count = await r
+        .knex('campaign_contact')
+        .where({ campaign_id: campaignId })
+        .count('*')
+
+      resp = { contacts: { count: parseInt(count[0].count) } }
+      res.writeHead(200, { 'Content-Type': 'application/json' })
+      res.end(JSON.stringify(resp))
+      break
+
+    case 'DELETE':
+      if (campaignaStatusShortCircuit(campaign, res)) {
+        break
```

Comment:
should this be `break` or `return`?  it looks like the the shortCircuit call already does the res.write().

Maybe we should just have something right at the top above the switch statement -- I guess this stops GET from reporting the data for archived campaigns, but I think that's ok.  It'll make the code simpler, and if we need contacts count by api for archived campaigns, we'll probably want a fuller analytics api endpoint anyway.

Reply:
👍  This inspired a bit of refactoring beyond the specific comments here.